src/plugins/expressions/common/ast/build_function.ts

```diff
@@ -0,0 +1,214 @@
+   * will not enforce whether multi-args are available, so only
+   * use this to update an existing arg if you are certain it
+   * is a multi-arg.
+   *
+   * @param name The name of the argument to add.
+   * @param value The value of the argument to add.
+   * @return `this`
+   */
+  addArgument: <A extends FunctionArgName<Fn>>(
+    name: A,
+    // TODO: Try to find a way to enforce multi-args in TS.
+    value: FunctionArgs<Fn>[A] | ExpressionAstExpressionBuilder
+  ) => this;
+  /**
+   * Retrieves an existing argument by name.
+   * Useful when you want to retrieve the current array of args and add
+   * something to it before calling `replaceArgument`.
+   *
+   * @param name The name of the argument to retrieve.
+   * @return `ExpressionAstArgument[]`
+   */
+  getArgument: <A extends FunctionArgName<Fn>>(name: A) => ExpressionAstArgument[] | undefined;
+  /**
+   * Overwrites an existing argument with a new value.
+   * In order to support multi-args, the value given must always be
+   * an array.
+   *
+   * @param name The name of the argument to replace.
+   * @param value The value of the argument. Must always be an array.
+   * @return `this`
+   */
+  replaceArgument: <A extends FunctionArgName<Fn>>(
+    name: A,
+    value: Array<FunctionArgs<Fn>[A]>
+  ) => this;
+  /**
+   * Removes an (optional) argument from the function.
+   *
+   * TypeScript will enforce that you only remove optional
+   * arguments. For manipulating required args, use `replaceArgument`.
+   *
+   * @param name The name of the argument to remove.
+   * @return `this`
+   */
+  removeArgument: <A extends OptionalKeys<FunctionArgs<Fn>>>(name: A) => this;
+  /**
+   * Converts function to an AST.
+   *
+   * @return `ExpressionAstFunction`
+   */
+  toAst: () => ExpressionAstFunction;
+  /**
+   * Converts function to an expression string.
+   *
+   * @return `string`
+   */
+  toString: () => string;
+}
+
+/**
+ * Manages an AST for a single expression function. The return value
+ * can be provided to `buildExpression` to add this function to an
+ * expression.
+ *
+ * Note that to preserve type safety and ensure no args are missing,
+ * all required arguments for the specified function must be provided
+ * up front. If desired, they can be changed or removed later.
+ *
+ * @param fnName String representing the name of this expression function.
+ * @param initialArgs Object containing the arguments to this function.
+ * @return `this`
+ */
+export function buildExpressionFunction<F extends string>(
+  fnName: F,
+  /**
+   * To support subexpressions, we override all args to also accept an
+   * ExpressionBuilder. This isn't perfectly typesafe since we don't
+   * know with certainty that the builder's output matches the required
+   * argument input, so we trust that folks using subexpressions in the
+   * builder know what they're doing.
+   */
+  initialArgs: { [K in keyof FunctionArgs<F>]: FunctionArgs<F>[K] | ExpressionAstExpressionBuilder }
```

Comment:
Highlighting this, as it's probably the main type-safety caveat:

The return value of `buildExpression` is added as an option for any expression function argument, so if someone is creating a subexpression they must take care to ensure that the expected return value of that whole expression actually satisfies the requirements of the argument it is passed to.

This is unfortunately not something we can determine for them in TS, as the output of an expression is equivalent to the output of the last function in that expression, which is something we don't know at compile time.